### PR TITLE
chore(engine): Rework null propagation

### PR DESCRIPTION
**Pull request by Finistere**
Previously for null propagation I was keeping track of the error path `["field", 0, "anotherFIeld"]` and re-traversing the graph at the end to mark the right fields as null on errors. It wasn't great, but it worked.

With `@inaccessible` it gets more tricky, at runtime we may encounter inaccessible enum values or objects behind an interface/union we need to hide from the user. In those cases we need to propagate a null without any error. The problem with the current implementation is that this will propagate null even though there might be fields used by the supergraph as requirements for subgraph requests. So need to hide data from the client, but not from the supergraph itself.

So I reworked null propagation to instead mark fields as inaccessible. The final serialization for the client will ignore those, but anything read by supergraph will work. I've also reworked the null propagation to simply keep track of the last nullable field within the error path. So now we have the `ErrorPath` for the error:

```rust
pub(crate) enum ErrorPathSegment {
    Field(SafeResponseKey),
    Index(usize),
    UnknownField(String),
}
```

and the one used during execution:

```rust
pub(crate) enum ResponseValueId {
    Field {
        object_id: ResponseObjectId,
        key: SafeResponseKey,
        nullable: bool,
    },
    Index {
        list_id: ResponseListId,
        index: u32,
        nullable: bool,
    },
}
```

The latter keeps track of more information while the former is only there for GraphQL errors. This allowed me to remove the `ResponseEdge` stuff with all the bitpacking.

This PR by itself doesn't solve the whole inaccessible issues as now I need to handle better how we behave when requirements aren't there. Today we just assumed them to be null, which doesn't work. Had to ignore a test today. I'm working on this, but this PR is already larger than it should...

Null propagation/add errors is still clunky in the code though, hope to improve it at some point. But the null propagation mechanism should be cleaner now.

### CODE REVIEW

File: crates/engine/src/response/write/merge.rs [hunk 1]
```diff
@@ -0,0 +1,96 @@
+use std::{cmp::Ordering, collections::VecDeque};
+
+use crate::response::{ResponseListId, ResponseObjectField, ResponseObjectId, ResponseValue};
+
+use super::ResponseBuilder;
+
+impl ResponseBuilder {
+    pub(super) fn recursive_merge_object(
```
Comment (Finistere) [inline crates/engine/src/response/write/merge.rs:8]: Moved code, same thing as before.

File: crates/engine/src/response/write/subgraph_response.rs [hunk 2]
```diff
@@ -0,0 +1,166 @@
+use std::{
```
Comment (Finistere) [inline crates/engine/src/response/write/subgraph_response.rs:1]: Moved code

File: crates/engine/src/execution/coordinator.rs [hunk 3]
```diff
@@ -503,45 +507,6 @@ impl<'ctx, R: Runtime> OperationExecution<'ctx, R> {
         (self, next_futures)
     }
 
-    fn get_first_key_and_default_object(
```
Comment (Finistere) [inline crates/engine/src/execution/coordinator.rs:506]: Moved that code

File: crates/engine/src/response/data.rs [hunk 4]
```diff
@@ -108,7 +124,89 @@ impl DataPart {
...
+    /// If we have an error for `name`, we still need to be able to read the `id`
+    /// field for the subgraph request retrieving `posts`. So we whichever field would be null
+    /// through propagation is marked as inaccessible. During serialization of the response to the
+    /// client those are treated as null. For every other purpose inaccessible fields are
+    /// transparent.
+    ///
+    /// The primary use-case where this kind of figure can happen where the supergraph needs to
+    /// propagate a null is for `@inaccessible` fields we detect at runtime. This happens typically
+    /// for inaccessible enum values or inaccessible objects we may encounter behind an
+    /// interface/union.
```
Comment (obmarg) [inline crates/engine/src/response/data.rs:166]: Good docstring

File: crates/engine/src/response/write/subgraph_response.rs [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+            return;
+        };
+        // we inverted the path.
+        let i = path.len() - i - 1;
+
+        // if let Some(value_id) = path.get(i).filter(|value_id| value_id.part_id() == self.data.id) {
+        //     self.data.make_inaccessible(*value_id);
+        // } else {
+        self.propagated_null_up_to_paths.push(path[..(i + 1)].to_vec());
+        // }
```
Comment (obmarg) [inline crates/engine/src/response/write/subgraph_response.rs:70]: What's with the commented out code?

**Review by obmarg — APPROVED** [review]

File: crates/engine/src/response/write/subgraph_response.rs (hunk 5 above)
Comment (Finistere) [reply crates/engine/src/response/write/subgraph_response.rs:70]: ah right forgot it, if the value is within the current part, I can just make the value inaccessible immediately. But somehow had an issue with it initially, so commented out to then fix further problems. bringing it back or giving up on it. thanks

**Pull request merged**